src/arithmetic/pattern_match.h

```diff
@@ -0,0 +1,667 @@
+/*!
+ *  Copyright (c) 2019 by Contributors
+ * \file tvm/arithmetic/pattern_match.h
+ *
+ * \brief Internal tool for expression-template based pattern matching.
+ *
+ * It helps to simplify pattern matching and rewrites.
+ * All the patterns are generated via expression template during compile time,
+ * so the result code should be as efficient as manually written pattern match code.
+ *
+ * The code below shows how to use the pattern matcher.
+ *
+ * \code
+ *
+ *  // max(x + z, y + z) => max(x, y) + z
+ *  arith::PVar<Expr> x, y, z;
+ *
+ *  // The following code tries to match the declared pattern.
+ *  // Match will fill the result of match into PVar if successful.
+ *  // Note that z occurs twice in the pattern,
+ *  // an equality check is performed to ensure each occurance of z
+ *  // is equivalent to each other.
+ *  if (max(x + z, y + z).Match(expr)) {
+ *    // Eval evaluates a pattern with the current matched value.
+ *    return (max(x, y) + z).Eval();
+ *  }
+ * \endcode
+ *
+ * \note The pattern matcher is not threadsafe,
+ *       do not use the same PVar in multiple threads.
```

Comment:
I guess saving a pattern into a variable may also be dangerous:
```C++
PVar<Expr> x, y, z;
auto pat = max(x + y, z);
pat.Match(expr); // x + y may be already destroyed here?
```
If it's indeed the case, it's worth mentioning here as well.

Reply:
We should confirm, my understanding is that it will be fine, because the temporary patterns will be destructed at the end of the current scope, so they are still valid here

Reply:
I particular, the above code will translate to

```c++
{
  PVar<Expr> x, y, z; 
  auto t0 = x + y;
  auto pat = max(t0, z);
  pat.Match(expr);
} // t0 is alive until here
```

Reply:
Also good point on the anti-pattern on bringing up the anti-pattern. I have updated comments to make people be aware of the potential misuse.

Reply:
I tried this code (if I'm not mistaken it's close to how the pattern classes work):
```C++
#include <iostream>

struct A {
    A() { std::cout << "A()\n"; }
    ~A() { std::cout << "~A()\n"; }
};

struct C {
    C(const A& a) : a_(a) { std::cout << "C()\n"; }
    ~C() { std::cout << "~C()\n"; }
    void match() const { std::cout << "match\n"; }
    const A& a_;
};

int main() {
    auto pat = C(A());
    pat.match();
}
```
The output is (both gcc and clang):
```
A()
C()
~A()
match
~C()
```
So A is destroyed before the call to match.

Reply:
Interesting, I didn't realize that was the case. Perhaps we need to do a deep construction by copy instead of by reference...